File: core/src/main/java/no/elg/infiniteBootleg/server/ClientBoundHandler.java
package no.elg.infiniteBootleg.server;

import static no.elg.infiniteBootleg.server.TowardsClientPacketsHandlerKt.handleClientBoundPackets;

import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;
import no.elg.infiniteBootleg.ClientMain;
import no.elg.infiniteBootleg.Main;
import no.elg.infiniteBootleg.protobuf.Packets.Packet;
import no.elg.infiniteBootleg.screens.ConnectingScreen;
import org.jetbrains.annotations.NotNull;

/** @author Elg */
public class ClientBoundHandler extends SimpleChannelInboundHandler<Packet> {

  public static final String TAG = "CLIENT";
  private final ServerClient client;

  public ClientBoundHandler(ServerClient client) {
    this.client = client;
  }

  @Override
  public void channelActive(@NotNull ChannelHandlerContext ctx) {
    client.ctx = ctx;
  }

  @Override
  public void channelInactive(@NotNull ChannelHandlerContext ctx) {
    Main.inst()
        .getScheduler()
        .executeSync(
            () -> {
              var serverClient = ClientMain.inst().getServerClient();
              if (serverClient != null) {
                var sharedInformation = serverClient.getSharedInformation();
                if (sharedInformation != null) {
                  var task = sharedInformation.getHeartbeatTask();
                  if (task != null) {
                    task.cancel(false);
                  }
                }
              }
              ClientMain.inst().setScreen(ConnectingScreen.INSTANCE);
            });
  }

  @Override
  protected void channelRead0(@NotNull ChannelHandlerContext ctx, @NotNull Packet packet) {
    //        Main.logger().log("Client bound packet " + packet.getType());
    if (packet.getDirection() == Packet.Direction.SERVER
        || packet.getType().name().startsWith("SB_")) {
      PacketExtraKt.fatal(
          ctx,
          "Client got a server packet " + packet.getType() + " direction " + packet.getDirection());
      return;
    }
    Main.inst().getScheduler().executeSync(() -> handleClientBoundPackets(client, packet));
  }

  @Override
  public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
    ConnectingScreen.INSTANCE.setInfo(
        "Exception caught, " + cause.getClass().getSimpleName() + ": " + cause.getMessage());
    cause.printStackTrace();
    ctx.close();
  }
}

File: core/src/main/kotlin/no/elg/infiniteBootleg/server/ClientBoundHandler.kt
package no.elg.infiniteBootleg.server

import io.netty.channel.ChannelHandlerContext
import io.netty.channel.SimpleChannelInboundHandler
import no.elg.infiniteBootleg.main.ClientMain
import no.elg.infiniteBootleg.main.Main
import no.elg.infiniteBootleg.protobuf.Packets
import no.elg.infiniteBootleg.screens.ConnectingScreen
import no.elg.infiniteBootleg.screens.ConnectingScreen.info

/**
 * @author Elg
 */
class ClientBoundHandler(private val client: ServerClient) : SimpleChannelInboundHandler<Packets.Packet>() {
  override fun channelActive(ctx: ChannelHandlerContext) {
    client.ctx = ChannelHandlerContextWrapper("client->server", ctx)
  }

  override fun channelInactive(ctx: ChannelHandlerContext) {
    Main.inst().scheduler.executeSync {
      val serverClient = ClientMain.inst().serverClient
      if (serverClient != null) {
        val sharedInformation = serverClient.sharedInformation
        if (sharedInformation != null) {
          val task = sharedInformation.heartbeatTask
          task?.cancel(false)
        }
      }
      ClientMain.inst().screen = ConnectingScreen
    }
  }

  override fun channelRead0(ctx: ChannelHandlerContext, packet: Packets.Packet) {
    if (packet.direction == Packets.Packet.Direction.SERVER || packet.type.name.startsWith("SB_")) {
      ctx.fatal("Client got a server packet ${packet.type} direction ${packet.direction}")
      return
    }
    client.handleClientBoundPackets(packet)
  }

  @Deprecated("Deprecated in Java")
  override fun exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
    info = "Exception caught, ${cause.javaClass.simpleName}: ${cause.message}"
    Main.logger().error(TAG, "Exception in netty IO", cause)
    ctx.close()
  }

  companion object {
    const val TAG = "CLIENT"
  }
}
